packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -180,12 +180,73 @@ class LineBreak extends TextBoundary {
   }
 }
 
+/// A text boundary that uses paragraphs as logical boundaries.
+///
+/// A paragraph is defined as the range between line terminators. If no
+/// line terminator exist then the paragraph boundary is the entire document.
+class ParagraphBoundary extends TextBoundary {
+  /// Creates a [ParagraphBoundary] with the text.
+  const ParagraphBoundary(this._text);
+
+  final String _text;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    return TextPosition(
+      offset: getTextBoundaryAt(position).start,
```

Comment:
nit: consider implementing these 2 separately. `getTextBoundaryAt` always searches for both boundaries instead of the one in the specified direction.